### ISSUE TITLE
Add support for qwen2.5-coder

### ISSUE THREAD

**Issue by kv-gits** [twinnydotdev/twinny#317]
Try https://ollama.com/library/qwen2.5-coder (1.5b, 7b) with FIM, but it doesnt work. It show in output some tokens e.g. <fim_suffix>. It would be nice to add special template for it.

**Comment by rjmacarthy** [comment]
Do you know which template it should be?

**Comment by kv-gits** [comment]
And what autocomplete (FIM) model is the best and tested with twinny?

**Comment by rjmacarthy** [comment]
I found that codellama-code or deepseek-base models work the best.

**Comment by kv-gits** [comment]
Thanks

**Comment by kirel** [comment]
I've experimented a bit after seeing the odd template of
```
❯ ollama show --template qwen2.5-coder:1.5b
{{- if .Suffix }}<|fim_prefix|>{{ .Prompt }}<|fim_suffix|>{{ .Suffix }}<|fim_middle|>
...
```
and the base models doesn't have much of a template. So I assumed you have to use the instruct model and you have to fill the `suffix` parameter in `api/generate` and lo and behold
```
❯ echo -e $(curl http://localhost:11434/api/generate -d '{
  "model": "qwen2.5-coder:1.5b",
  "prompt": "def fib(", "suffix": " "
}' | jq -s 'map(.response) | join("")')
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  6223    0  6149  100    74   4721     56  0:00:01  0:00:01 --:--:--  4775
"n):
 \"\"\"
 Compute the nth Fibonacci number.

 Args:
 n (int): The position in the Fibonacci sequence to compute.

 Returns:
 int: The nth Fibonacci number.
 \"\"\"
 if n <= 1:
 return n
 "
```
while when leaving out the suffix or even setting it to empty string I get
```
❯ echo -e $(curl http://localhost:11434/api/generate -d '{
  "model": "qwen2.5-coder:1.5b",
  "prompt": "def fib(", "suffix": ""
}' | jq -s 'map(.response) | join("")')
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 26463    0 26390  100    73   6107     16  0:00:04  0:00:04 --:--:--  6125
" It looks like you're trying to define a function named `fib` in Python. However, there seems to be an issue with the syntax of your code. The `fib` function should be properly defined using parentheses and arguments.

Here's a corrected version of your code:

```python
def fib(n):
 # Base cases for the first two Fibonacci numbers
 if n == 0:
 return 0
 elif n == 1:
 return 1

 # Recursive case to calculate Fibonacci sequence
 else:
 return fib(n-1) + fib(n-2)
```

In this code:
- The function `fib` takes one argument `n`.
- The base cases handle the first two Fibonacci numbers (`0` and `1`).
- The recursive case calculates the Fibonacci number by summing the two preceding ones.

You can call this function with an integer argument to compute the corresponding Fibonacci number. For example:

```python
print(fib(5)) # Output: 8 (since 5th Fibonacci number is 8)
```

This will print `8`, which is the fifth Fibonacci number in the sequence."
```

which is not what we want.

Do you have any guidance on how to make this work with Twinny?

**Comment by kirel** [comment]
I can get similar results with
```
echo -e $(curl http://localhost:11434/api/generate -d '{
  "model": "qwen2.5-coder:1.5b",
  "prompt": "def fib(", "template": "<|fim_prefix|>{{ .Prompt }}<|fim_suffix|>{{ .Suffix }}<|fim_middle|>"
}' | jq -s 'map(.response) | join("")')
```
by overriding the template

**Comment by kirel** [comment]
Ok for future googlers - my workaround for now is:

```
cat <<EOF > modelfile
FROM qwen2.5-coder:1.5b
TEMPLATE "{{ .Prompt }}"
EOF
ollama create qwen2.5-coder:1.5b-plain -f modelfile
rm modelfile
echo -e $(curl http://localhost:11434/api/generate -d '{
  "model": "qwen2.5-coder:1.5b-plain",
  "prompt": "<|fim_prefix|>def fib(<|fim_suffix|><|fim_middle|>"
}' | jq -s 'map(.response) | join("")')
```

set the custom fim template to
```
<|fim_prefix|>{{{prefix}}}<|fim_suffix|>{{{suffix}}}<|fim_middle|>
```
and Twinny completes like a champ when selecting `qwen2.5-coder:1.5b-plain` model.
Go for `7b` if you can.

**Comment by rjmacarthy** [comment]
Hey @kirel is it possible to get the same result by using the deepseek fim template? Seems maybe just the model needs to be `plain`? Thanks.

**Comment by kirel** [comment]
There qwen models are trained on the template above. There is no reason to use a different one. The problem is that with ollama's default twinny isn't passing it so that it's compatible.

**Comment by rjmacarthy** [comment]
https://github.com/twinnydotdev/twinny/blob/e5d2f6ceafa15affb24419238c3cf303dd26d413/src/extension/fim-templates.ts#L96 

Thanks, sorry I thought deepseek's template was the same. Seems we already have this template declared? Will this work?

**Comment by rjmacarthy** [comment]
Thanks @kirel.  FYI Further to this I just released v3.17.11 which adds proper support for codeqwen2.5.  There is no need to create a new model using a model file as suggested and you can just use the base model e.g `qwen2.5-coder:7b-base`, it works perfectly fine.  The issue was actually the extra context which was added to the suffix in the original template. @kv-gits if you now choose this model with the codeqwen fim template all should work perfectly.  Many thanks.

**Comment by kirel** [comment]
Cool, thanks! No dirty workarounds 🤝
